### PR TITLE
nitdoc: Use wikilinks on README files

### PR DESCRIPTION
**Pull request by Morriar**
This PR introduce the use of wikilinks in README files.

It's mainly a request for comments on what you want, what you need and what you like from Nitdoc and README writting.

0f6ff78 gives an example of README file written using wikilinks. The processed result by nitdoc is visible [here](http://gresil.org/jenkins/job/CI-nitdoc/ws/doc/stdlib/group_github.html).

Supported commands for now:

* `[[doc:MEntity]]`: include the documentation article or an MEntity
* `[[list:MEntity]]`: list classes or methods from than MEntity

Old API pages still available, there are prefixed with `api_`.

Demos from [Jenkins::CI-nitdoc](http://gresil.org/jenkins/job/CI-nitdoc):
* [Standard library](http://gresil.org/jenkins/job/CI-nitdoc/ws/doc/stdlib/index.html)
* [Nit compilers and tools](http://gresil.org/jenkins/job/CI-nitdoc/ws/doc/nitc/index.html)

Some question still need to be answered:
* What to do with old API page (http://gresil.org/jenkins/job/CI-nitdoc/ws/doc/stdlib/api_group_github.html)
* What comment to display in API page (http://gresil.org/jenkins/job/CI-nitdoc/ws/doc/stdlib/api_group_github.html)
* What commands should be supported by wikilinks directive?

### CODE REVIEW

File: src/doc/doc_commands.nit [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+		else if command_string.has_prefix("code:") then
+			return new CodeCommand(command_string)
+		end
+		return new UnknownCommand(command_string)
+	end
+
+	redef fun to_s do return string
+end
+
+# Used to factoryse initialization of DocCommands.
```
Comment (R4PaSs) [inline src/doc/doc_commands.nit]: :s/factoryse/factorize

File: src/doc/doc_phases/doc_readme.nit [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+		v.open_article
+	end
+
+	redef fun add_wikilink(v, token) do
+		var link = token.link.to_s
+		var cmd = new DocCommand(link)
+		if cmd isa UnknownCommand then
+			# search MEntities by name
+			var res = v.phase.doc.mentities_by_name(link.to_s)
+			# no match, print warning and display wikilink as this
```
Comment (R4PaSs) [inline src/doc/doc_phases/doc_readme.nit]: :s/as this/as is

**Comment by R4PaSs** [comment]
This might be me, but I'm a bit surprised by the Event::list in the Github API README, the events listed are those from SDL, maybe there was some kind of dependancy at one point (or maybe it's a bug in wikilinks ?), however it seems a bit strange to my eyes right now.

Other than that, I like the way information is presented, it is simple, clear, and most of the useful information is conveyed in a simple README, Nitdoc does a really good job at presenting all this, kudos to you @Morriar !

On the project view then, is it still necessary to print information ? If we take a look at the Github group, the readme is not rendered using wikilinks, and the Markdown is printed as-is, which give little to no information on the project itself.
Should we print a short version instead ? The `github.nit` module header maybe ?

For your questions now:

* What to do with old API page : The API page is still useful I think, but as a catalog of the classes and services available in them. Some information on the structure of the class should be displayed imho.
The group page could be useful for newcomers as it gives a quick oversight of the library, but keeps it to a strict minimum for a quick grasp of the `how to use` part.
I see the API page as someplace to go into the details of the group, eventually in order to extend it, which requires a bit more info on the internals.
Maybe we should add a link from the group page to the API to access it ?
* What commands should be supported by wikilinks directive? I'm not sure I know what my options are here, can you elaborate on this point so I can give you an answer which will, hopefully, not be completely off-topic ?

On the PR, lots of code, but aside from the few typos I noticed, :shipit: !

**Comment by Morriar** [comment]
Reroll with @R4PaSs remarks

**Comment by Morriar** [comment]
@R4PaSs 

* I fixed the name conflict in the readme file.

This makes me think of another question:

Nitdoc displays warning when a MEntity name is not found:
~~~sh
lib/github/README.md:5,1: cannot find article for `LOL` (readme-warning)
	[[doc: LOL]]
	^
~~~

Same thing for the conflicts (like with `[[doc: events]]`):
~~~sh
lib/github/README.md:69,1: conflicting article for `events` (choices : sdl::events, api::events, api#Issue#events) (readme-warning)
	[[list: events]]
	^
~~~

The problem with conflicts it's they occur really often with classes. Example with [[doc: GithubAPI]] do you mean the intro in api::GithubAPI or one of the redef?

For now, nitdoc looks at all MEntities with the name `GithubAPI` and returns 1 MClass and 2 MClassdefs (one intro and one redef). It selects the first result, the MClass (because of the implementation right now) adds it to the page then displays a warning about the two other results

Users can already desambiguise with the namespace notation: `[[doc: cache::GithubAPI]]` to turn off the warning. But it's stupid to have to write [[doc: api::GithubAPI]] when you can write [[doc: GithubAPI]] and it fits 99% of the use cases.
So I'm thinking to relax this rule by looking at the MClass first and let the user desambiguise with the namespace without displaying a warning.

**Comment by Morriar** [comment]
Other problem that I see is when displaying the README in github.
Wikilinks that include things like `[[list:]]` or `[[doc:]]` are not that helpfull when reading plain markdown.

**Comment by privat** [comment]
on the index, the readme seems to be unprocessed http://gresil.org/jenkins/job/CI-nitdoc/ws/doc/stdlib/index.html#github.definition

**Comment by privat** [comment]
Looks very great. I plan to try to use it once it is merged. +1

For the various questions, I think that you should just chose something I you did, then discuss and solve each independent issue with future PRs. You can create a TODO issue with check boxes to lise them (because they will be hard to track back in this PR)

**Comment by Morriar** [comment]
Reroll:

* no more blank page for empty readme (use the API page instead)
* no more unprocessed wikilinks in Nitdoc API pages (replaced with synopsis from README)
  * this implies that the first line of a README never contains a wikilinks that include something
* added a link to swtich between API and README
* produce warnings for empty README

There stills a lot of work but I think that this PR is ready to be merged (if tests are ok). The demo should be updated soon.

**Comment by privat** [comment]
This one conflict with the soon merged #1446, so I assume that I should wait before merging this one.

**Comment by privat** [comment]
Tried to resolve merge conflicts myself. It didn't work, so I let you hold the scalpel.

**Comment by Morriar** [comment]
Reroll rebased on master

**Comment by privat** [comment]
~~~
$ git merge origin/master
$ make
$ make docs
[...]
Runtime error: Uninitialized attribute _markdown_processor (doc/html_templates/html_templates.nit:582)
~~~

cf http://gresil.org/jenkins/job/nit_test_tools/737/consoleFull

**Comment by Morriar** [comment]
@privat rerolled without the crash on `mardown_processor`

**Pull request merged**